File: 0014_longest-common-prefix/solution.go
package _014_longest_common_prefix


// 横向扫描, 依次遍历字符串数组中的每个字符串，对于每个遍历到的字符串，更新最长公共前缀，当遍历完所有的字符串以后，即可得到字符串数组中的最长公共前缀。
// 如果在尚未遍历完所有的字符串时，最长公共前缀已经是空串，则最长公共前缀一定是空串，因此不需要继续遍历剩下的字符串，直接返回空串即可。
func longestCommonPrefix(strs []string) string {
    if len(strs) == 0 {
        return ""
    }
    prefix := strs[0]
    count := len(strs)
    for i := 1; i < count; i++ {
        prefix = lcp(prefix, strs[i])
        if len(prefix) == 0 {
            break
        }
    }
    return prefix
}

func lcp(str1, str2 string) string {
    length := min(len(str1), len(str2))
    index := 0
    for index < length && str1[index] == str2[index] {
        index++
    }
    return str1[:index]
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}

File: 0014_longest-common-prefix/solution.py
from typing import List


class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        """纵向扫描
            停止扫描的情况：
            1.最长公共前缀的最大长度<= 数组中最短字符串长度
            2.按列比较，值不相等，则停止

        """
        if not strs:
            return ""
        length, count = len(strs[0]), len(strs)
        for i in range(length):
            c = strs[0][i]
            for j in range(1, count):
                if len(strs[j]) == i or c != strs[j][i]:
                    return strs[0][:i]
        return strs[0]
